### ISSUE TITLE
Is it possible to finetune the pretrained model on casual language modeling or text summarization?

### ISSUE THREAD

**Issue by fabrahman** [allenai/longformer#29]
Hi,
Thanks for providing and presenting this nice work.

As mentioned in your paper, your attention pattern for modeling long sequences can be plugged into any pretrained transformer model.
I wonder if this repo covers code to finetune a pretrained LM (e.g. gpt-2) or your own released pretrained model on a new dataset for language modeling task?

If so, is it possible through PyTorch implementation or the CUDA kernel?
I would appreciate if you can guide me in this respect.

**Comment by ibeltagy** [comment]
We didn't release the pretraining code, but you can probably use the [HF's one](https://github.com/huggingface/transformers/blob/master/examples/run_language_modeling.py). You can also start with the trick of copying the position embeddings (discussed at the beginning of section 5) with no pretraining at all, which usually works reasonably well especially if your downstream supervised data is large. 

Casual language modeling: it is available with the `autoregressive=True` flag in the CUDA kernel, and you can use it to pretrain gpt-2. Notice that `autoregressive=True` only works for the local attention, you still need some masking for the global attention (in case you decided to use it). 

Text summarization: generative or extractive? I would expect extractive summarization to work easily with our pretrained Longformer. For generative summarization, you will need your pretrained gpt-2, or your pretrained BART. Using the LongfomrerSelfAttention for BART is going to be interesting because it has two selfattention blocks.

**Comment by Hannabrahman** [comment]
> We didn't release the pretraining code, but you can probably use the [HF's one](https://github.com/huggingface/transformers/blob/master/examples/run_language_modeling.py). You can also start with the trick of copying the position embeddings (discussed at the beginning of section 5) with no pretraining at all, which usually works reasonably well especially if your downstream supervised data is large.
> 
> Casual language modeling: it is available with the `autoregressive=True` flag in the CUDA kernel, and you can use it to pretrain gpt-2. Notice that `autoregressive=True` only works for the local attention, you still need some masking for the global attention (in case you decided to use it).
> 
> Text summarization: generative or extractive? I would expect extractive summarization to work easily with our pretrained Longformer. For generative summarization, you will need your pretrained gpt-2, or your pretrained BART. Using the LongfomrerSelfAttention for BART is going to be interesting because it has two selfattention blocks.

Thanks for quick reply.
I have tried finetuning GPT-2 (bye HF), however my dataset input sequences exceeds the gpt-2 sequence length limitation and I had to truncate it which results in loosing a large amount of information.

Would you mind including some instruction of how to use the CUDA kernel for LM?
Also, when you said extractive summarization is possible through your pretrained Longformer, is it again on CUDA kernel?

Sorry, I've got confused about the "How to use" section in readme is for PyTorch implementation, right?

**Comment by ibeltagy** [comment]
>  exceeds the gpt-2 sequence length limitation
sure, you need to modify the model a bit to increase the size of the position embedding (that's where the copy trick I mentioned is useful)

> use the CUDA kernel for LM
The `LongformerSelfattention` class is a drop-in replacement of the `BertSelfattention` class. You need to edit the gpt-2 code to replace its selfattention with longformer's

> How to use
It is referring to the use of the pretrained Longformer, which is not what you want. You want to plug `LongformerSelfattention` to be used inside gpt-2

**Comment by ohmeow** [comment]
> For generative summarization, you will need your pretrained gpt-2, or your pretrained BART. Using the LongfomrerSelfAttention for BART is going to be interesting because it has two selfattention blocks.

W/r/t BART, how would you handle the cross-encoder-attention used by the decoder given that LSA actually has two attention mechanisms?  And in the case of summarization specifically, would it make sense to just use LSA in the encoder since the summaries are small by nature?

Would love to extend my BART model from 1024 to over 4K tokens, so really interested to understand how this could be done.

Thanks.

**Comment by ibeltagy** [comment]
@ohmeow, please check the answer here https://github.com/allenai/longformer/issues/28#issuecomment-635423494. As you said, if you assume that the output length is short, you can use `LongformerSelfAttention` for the encoder and continue using the regular n^2 selfattention for the two selfattention components in the decoder. I think it will be possible to run on inputs of length up to 16k tokens.

**Comment by ibeltagy** [comment]
Closing. Please feel free to reopen or create a new one if you have more questions.

**Comment by santhoshkolloju** [comment]
@ibeltagy
Thanks for the good work on longformer , 
How to i make the longformerAttention to use a unidirectional attention and not bidirectional as in the case of Roberta?Can you please add notebook with GPT-2 as example.

**Comment by chandu7077** [comment]
How to use longformer for extractive summarization?

**Comment by ibeltagy** [comment]
@santhoshkolloju , good question. You can actually do that without any masks but will require some changes in the attention implementation. Check [here](https://github.com/allenai/longformer/blob/master/longformer/sliding_chunks.py#L73-L79) for the code that computes attention scores. You want to keep the main diagonal and the lower triangle and skip the upper triangle. [This function](https://github.com/allenai/longformer/blob/master/longformer/sliding_chunks.py#L88) will need to change slightly as well. 

Similarly if you use the `'sliding_chunks_no_overlap'` implementation, you will delete one of [these three lines](https://github.com/allenai/longformer/blob/master/longformer/sliding_chunks.py#L158-L160) that corresponds to the upper triangle and update [this](https://github.com/allenai/longformer/blob/master/longformer/sliding_chunks.py#L166) accordingly.

**Comment by ibeltagy** [comment]
@chandu7077, you would use it the same way you would use any other pretrained model, formulate the problem as a binary classification problem for sentences or a sequence labeling for spans. You might want to experiment with global attention and find out which configuration works for your setting. One potential configuration is having separator tokens sentences, use them for classification and give them global attention.

**Comment by Aakash12980** [comment]
Is Bart computationally heavy than Bert? I am using Colab which provide 12 Gb of GPU storage. Also I am planning to use Bart for sentence simplification with 130k training data. Previously I used Bert2Bert EncoderDecoderModel from HG but the result is poor after training the model for 4 epochs. Should I go for Bart? The project is for my undergrad project.
